flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -148,11 +148,7 @@
 
 	// ------------------------------------------------------------------------
 
-	/**
-	 * All interaction with the {@code StreamOperator} must be synchronized on this lock object to
-	 * ensure that we don't have concurrent method calls that void consistent checkpoints.
-	 */
-	private final Object lock = new Object();
+	private final ExecutionDecorator.SynchronizedExecutionDecorator executionDecorator;
```

Comment:
Why do we need the subtype here instead of just the decorator?

Reply:
SynchronizedExecutionDecorator gives access to its lock object.

Reply:
I'd rather keep lock separate and inject it into the executor.

Reply:
Why? It spreads the responsibility of managing the lock.

Reply:
I'd value proper abstraction through interfaces + the ability to inject dependencies over the separation of concerns. For me, the usage of a concrete class over an interface, especially on field level, is a big smell.
Btw I don't see any "management" overhead as the lifecycle of the lock is then bound to the StreamTask.

Reply:
It's going to be interface after the `getCheckpointLock()` method will be removed - hopefully, next PR.